optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -180,7 +190,332 @@
+                    indexOfFirstFilter = i;
+                }
+            } else if (hasAFilter) {
+                throw new IllegalStateException("Indexing joiner (" + joiner + ") must not follow a filtering joiner ("
+                        + joiners[indexOfFirstFilter] + ").");
+            }
+        }
+        if (indexOfFirstFilter < 0) { // Only found indexing joiners.
+            return join(otherClass, AbstractBiJoiner.merge(joiners));
+        }
+        // Assemble the join stream.
+        BiConstraintStream<A, B> joined = indexOfFirstFilter == 0 ?
+                join(otherClass) :
+                join(otherClass, Arrays.copyOf(joiners, indexOfFirstFilter));
+        // We merge all filters into one, so that we don't pay the penalty for lack of indexing more than once.
+        BiPredicate<A, B> resultingFilter = (a, b) -> true;
+        for (int i = indexOfFirstFilter; i < joinerCount; i++) {
+            FilteringBiJoiner<A, B> filteringJoiner = (FilteringBiJoiner<A, B>) joiners[i];
+            resultingFilter = resultingFilter.and(filteringJoiner.getFilter());
+        }
+        return joined.filter(resultingFilter);
+    }
+
+    /**
+     * Create a new {@link BiConstraintStream} for every two different A.
+     * <p>
+     * Important: This is faster and more scalable than a {@link #joinOther(UniConstraintStream) joinOther}
+     * followed by a {@link BiConstraintStream#filter(BiPredicate) filter},
+     * because it applies hashing and/or indexing on the properties,
+     * so it doesn't create nor checks every combination of A and B.
+     * @param otherStream never null
+     * @return a stream that matches every combination of A and a different A
+     */
+    default BiConstraintStream<A, A> joinOther(UniConstraintStream<A> otherStream) {
+        return join(otherStream, Joiners.notEqual());
+    }
+
+    /**
+     * Create a new {@link BiConstraintStream} for every combination of A and a different A for which the
+     * {@link BiJoiner} is true (for the properties it extracts from both facts).
+     * <p>
+     * Important: This is faster and more scalable than a {@link #joinOther(UniConstraintStream) joinOther}
+     * followed by a {@link BiConstraintStream#filter(BiPredicate) filter},
+     * because it applies hashing and/or indexing on the properties,
+     * so it doesn't create nor checks every combination of A and B.
+     * @param otherStream never null
+     * @param joiner never null
+     * @return a stream that matches every combination of A and a different A for which the {@link BiJoiner} is true
+     */
+    default BiConstraintStream<A, A> joinOther(UniConstraintStream<A> otherStream, BiJoiner<A, A> joiner) {
+        if (joiner instanceof NoneBiJoiner) {
+            return join(otherStream, Joiners.notEqual());
+        } else {
+            return join(otherStream, AbstractBiJoiner.merge(Joiners.notEqual(), joiner));
```

Comment:
If nonEquals is indexed, does this merge make sure that if we do:

`joinOther(Foo.class, equal(getType), filtering(a, b -> a.getAge * b.getAge == 42))`

it is syntactic sugar for:

`join(Foo.class, equal(getType), notEqual(), filtering(a, b -> a.getAge * b.getAge == 42))`

So the first joiner remained the first joiner, but the second joiner became third joiner, because it's the first non-indexed joiner.

Reply:
Ordering of joiners is never changed. When `notEqual()` is added by the `joinOther()` statement, it as added as *the first* joiner.